### PR TITLE
Initial SPEL code (Interpreter only)

### PR DESCRIPTION
**Pull request by dtillman**

### CODE REVIEW

File: src/Common/src/Common.Expression/Steeltoe.Common.Expression.csproj [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  <Import Project="..\..\..\..\versions.props" />
+  <Import Project="..\..\..\..\sharedproject.props" />
+
+  <ItemGroup>
+    <ProjectReference Include="..\Common\Steeltoe.Common.csproj" />
+  </ItemGroup>
+  
+  <ItemGroup>
+    <PackageReference Include="System.Reflection.Emit.Lightweight" Version="4.7.0" />
```
Comment (hananiel) [inline src/Common/src/Common.Expression/Steeltoe.Common.Expression.csproj:18]: Should we add to versions.props and use a variable?

**Review by hananiel — COMMENTED** [review]
A few places had commented out code - some are obvious as to why ex. compiled mode is not yet implemented. Others like proxy - it might be helpful to add a comment or TODO as to what the next action is - in case someone else attempts it.

File: src/Common/src/Common.Expression/Spring/CodeFlow.cs [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+        }
+
+        public static string ToDescriptor(Type type)
+        {
+            if (type.IsPrimitive)
+            {
+                var name = type.Name;
+                switch (name.Length)
+                {
+                    case 4:
```
[the remark this answers is not in the document]
Comment (dtillman) [reply src/Common/src/Common.Expression/Spring/CodeFlow.cs:28]: I'm guessing for perf .. for the most part I just ported the java code pretty much as is initially to get things working and all the unit tests to pass.

File: src/Common/src/Common.Expression/Steeltoe.Common.Expression.csproj (hunk 1 above)
Comment (dtillman) [reply src/Common/src/Common.Expression/Steeltoe.Common.Expression.csproj:18]: Ya .. good point... will do.

Comment (dtillman) [reply src/Common/src/Common.Expression/Steeltoe.Common.Expression.csproj]: I'll look at the TODOs .. several of them are from the Java code .. I just copied them over as part of comments.. I'll update the ones I added

Comment (hananiel) [reply src/Common/src/Common.Expression/Steeltoe.Common.Expression.csproj]: Ok once you are done making any more changes, let me know here and I will approve and we can get this merged in.

File: src/Common/test/Common.Expression.Test/Spring/TestResources/TestPerson.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        {
+            get => name;
+            set => name = value;
+        }
+
+        public TestAddress Address
+        {
+            get => address;
+            set => address = value;
+        }
```
Comment (TimHess) [inline src/Common/test/Common.Expression.Test/Spring/TestResources/TestPerson.cs:22]:
```suggestion
        public string Name { get; set; }

        public TestAddress Address { get; set; }
```

File: src/Common/src/Common/Expression/CSharp/ExpressionParser.cs [hunk 4]
```diff
@@ -1,21 +0,0 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the Apache 2.0 License.
-// See the LICENSE file in the project root for more information.
-
-using System;
-
-namespace Steeltoe.Common.Expression.CSharp
-{
-    public class ExpressionParser : IExpressionParser
```
Comment (TimHess) [inline src/Common/src/Common/Expression/CSharp/ExpressionParser.cs:9]: Is a CSharp expression parser something we are still planning to add in the future? I don't see it in the backlog

File: src/Common/src/Common.Expression/TypedValue.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        public Type TypeDescriptor { get; }
+
+        public override bool Equals(object other)
+        {
+            if (this == other)
+            {
+                return true;
+            }
+
+            if (!(other is TypedValue))
```
Comment (TimHess) [inline src/Common/src/Common.Expression/TypedValue.cs:37]:
```suggestion
            if (other is not TypedValue)
```

File: src/Common/src/Abstractions/Expression/IServiceResolver.cs [hunk 6]
```diff
@@ -2,12 +2,14 @@
...
+using System.Collections.Generic;
+using System.Text;
 
-namespace Steeltoe.Messaging.RabbitMQ.Expressions
+namespace Steeltoe.Common.Expression
 {
     public interface IServiceResolver
     {
-        object Resolve(IEvaluationContext context, string serviceName);
+        object Resolve(IEvaluationContext context, string beanName);
```
Comment (TimHess) [inline src/Common/src/Abstractions/Expression/IServiceResolver.cs:13]: This doesn't appear to be Spring-specific, should it stay as `serviceName`?

File: src/Common/test/Common.Expression.Test/Spring/TestResources/TestPerson.cs (hunk 3 above)
Comment (dtillman) [reply src/Common/test/Common.Expression.Test/Spring/TestResources/TestPerson.cs:22]: This change will likely break a bunch of unit tests due to the way they test reading and writing properties vs fields... so I'm going to leave alone for now

File: src/Common/src/Common/Expression/CSharp/ExpressionParser.cs (hunk 4 above)
Comment (dtillman) [reply src/Common/src/Common/Expression/CSharp/ExpressionParser.cs:9]: Not at this point

File: src/Common/src/Abstractions/Expression/IServiceResolver.cs (hunk 6 above)
Comment (dtillman) [reply src/Common/src/Abstractions/Expression/IServiceResolver.cs:13]: Ya ... it should stay serviceName

File: src/Common/src/Abstractions/Expression/IServiceResolver.cs [hunk 7]
```diff
@@ -2,12 +2,14 @@
...
+using System.Collections.Generic;
+using System.Text;
 
-namespace Steeltoe.Messaging.RabbitMQ.Expressions
+namespace Steeltoe.Common.Expression.Internal
 {
     public interface IServiceResolver
     {
-        object Resolve(IEvaluationContext context, string serviceName);
+        object Resolve(IEvaluationContext context, string beanName);
```
Comment (TimHess) [inline src/Common/src/Abstractions/Expression/IServiceResolver.cs:13]:
```suggestion
        object Resolve(IEvaluationContext context, string serviceName);
```

File: src/Common/src/Common.Expression/Spring/Ast/FunctionReference.cs [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+        public override ITypedValue GetValueInternal(ExpressionState state)
+        {
+            var value = state.LookupVariable(_name);
+            if (value == TypedValue.NULL)
+            {
+                throw new SpelEvaluationException(StartPosition, SpelMessage.FUNCTION_NOT_DEFINED, _name);
+            }
+
+            if (!(value.Value is MethodInfo))
```
Comment (TimHess) [inline src/Common/src/Common.Expression/Spring/Ast/FunctionReference.cs:37]:
```suggestion
            if (value.Value is not MethodInfo)
```

File: src/Common/src/Common.Expression/Spring/Ast/FormatHelper.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+                    items.Add(FormatClassNameForMessage(null));
+                }
+            }
+
+            return name + "(" + string.Join(",", items) + ")";
+        }
+
+        public static string FormatClassNameForMessage(Type clazz)
+        {
+            return clazz != null ? clazz.FullName : "null";
```
Comment (TimHess) [inline src/Common/src/Common.Expression/Spring/Ast/FormatHelper.cs:33]:
```suggestion
            return clazz?.FullName;
```

File: src/Common/src/Common.Expression/Spring/Ast/FunctionReference.cs [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+            {
+                functionArgs = ReflectionHelper.SetupArgumentsForVarargsInvocation(ClassUtils.GetParameterTypes(method), functionArgs);
+            }
+
+            var compilable = false;
+            try
+            {
+                var result = method.Invoke(method.GetType(), functionArgs);
+                compilable = !argumentConversionOccurred;
+                return new TypedValue(result, result == null ? method.ReturnType : result.GetType());
```
Comment (TimHess) [inline src/Common/src/Common.Expression/Spring/Ast/FunctionReference.cs:142]:
```suggestion
                return new TypedValue(result, result?.GetType() ?? method.ReturnType);
```

File: src/Common/src/Common.Expression/Spring/Support/ReflectiveMethodExecutor.cs [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+            {
+                _argumentConversionOccurred = ReflectionHelper.ConvertArguments(context.TypeConverter, arguments, _originalMethod, _varargsPosition);
+                if (_originalMethod.IsVarArgs())
+                {
+                    arguments = ReflectionHelper.SetupArgumentsForVarargsInvocation(ClassUtils.GetParameterTypes(_originalMethod), arguments);
+                }
+
+                // ReflectionUtils.makeAccessible(this.methodToInvoke);
+                var value = _methodToInvoke.Invoke(target, arguments);
+                return new TypedValue(value, value == null ? _originalMethod.ReturnType : value.GetType()); // new TypeDescriptor(new MethodParameter(this.originalMethod, -1)).Narrow(value));
```
Comment (TimHess) [inline src/Common/src/Common.Expression/Spring/Support/ReflectiveMethodExecutor.cs:91]:
```suggestion
                return new TypedValue(value, value?.GetType() ?? _originalMethod.ReturnType); // new TypeDescriptor(new MethodParameter(this.originalMethod, -1)).Narrow(value));
```

File: src/Common/src/Common.Expression/TypedValue.cs (hunk 5 above)
Comment (TimHess) [reply src/Common/src/Common.Expression/TypedValue.cs:37]: This is for cosmetics/readability, it is not the only case and is completely optional

File: src/Common/src/Common.Expression/Spring/Ast/FormatHelper.cs (hunk 9 above)
Comment (TimHess) [reply src/Common/src/Common.Expression/Spring/Ast/FormatHelper.cs:33]: This is for cosmetics/readability, it is not the only case and is completely optional

File: src/Common/src/Common.Expression/Spring/Ast/FunctionReference.cs (hunk 10 above)
Comment (TimHess) [reply src/Common/src/Common.Expression/Spring/Ast/FunctionReference.cs:142]: This is for cosmetics/readability, it is not the only case and is completely optional

File: src/Common/src/Common.Expression/Spring/Ast/OpDec.cs [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+            var lvalue = operand.GetValueRef(state);
+
+            var operandTypedValue = lvalue.GetValue();  // operand.getValueInternal(state);
+            var operandValue = operandTypedValue.Value;
+            var returnValue = operandTypedValue;
+            ITypedValue newValue = null;
+
+            if (IsNumber(operandValue))
+            {
+                if (operandValue is decimal)
```
Comment (TimHess) [inline src/Common/src/Common.Expression/Spring/Ast/OpDec.cs:39]: I think a [pattern matching switch statement](https://docs.microsoft.com/en-us/dotnet/csharp/pattern-matching#using-pattern-matching-switch-statements) would improve readability here

File: src/Common/src/Common.Expression/Spring/Ast/ServiceReference.cs [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+        public ServiceReference(int startPos, int endPos, string serviceName)
+            : base(startPos, endPos)
+        {
+            _serviceName = serviceName;
+        }
+
+        public override ITypedValue GetValueInternal(ExpressionState expressionState)
+        {
+            var beanResolver = expressionState.EvaluationContext.ServiceResolver;
```
Comment (TimHess) [inline src/Common/src/Common.Expression/Spring/Ast/ServiceReference.cs:23]: serviceResolver?

**Review by TimHess — APPROVED** [review]
My comments/suggestions are all cosmetic and not required. There are more instances where "bean" could be changed to "service" (total of 55 in `src/` and 250 in `test/`) and pattern matching could be used, but the changes aren't required. Happy to re-approve if you want to push those changes

File: src/Common/src/Common.Expression/Spring/Ast/OpDec.cs (hunk 12 above)
Comment (dtillman) [reply src/Common/src/Common.Expression/Spring/Ast/OpDec.cs:39]: I'll update on next PR that finishes this and cleans up commented out code ... etc.

File: src/Common/src/Common.Expression/Spring/Ast/ServiceReference.cs (hunk 13 above)
Comment (dtillman) [reply src/Common/src/Common.Expression/Spring/Ast/ServiceReference.cs:23]: Same with this one

**Review by hananiel — APPROVED** [review]

**Pull request merged**